internal/pkg/kmsg/kmsg.go

```diff
@@ -12,73 +12,37 @@ import (
 	"io"
 	"log"
 	"os"
-	"path/filepath"
 
 	"golang.org/x/sys/unix"
-
-	"github.com/talos-systems/talos/pkg/constants"
 )
 
-// Setup configures the log package to write to the kernel ring buffer via
-// /dev/kmsg.
-func Setup(prefix string, withLogFile bool) error {
-	kmsg, err := os.OpenFile("/dev/kmsg", os.O_RDWR|unix.O_CLOEXEC|unix.O_NONBLOCK|unix.O_NOCTTY, 0666)
-	if err != nil {
-		return fmt.Errorf("failed to open /dev/kmsg: %w", err)
-	}
-
-	var writer io.Writer = &Writer{KmsgWriter: kmsg}
-
-	if withLogFile {
-		if err := os.MkdirAll(constants.DefaultLogPath, 0700); err != nil {
-			return err
-		}
-
-		logPath := filepath.Join(constants.DefaultLogPath, "machined.log")
-
-		f, err := os.OpenFile(logPath, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
-		if err != nil {
-			return fmt.Errorf("failed to open %s: %w", logPath, err)
-		}
-
-		writer = io.MultiWriter(writer, f)
-	}
-
-	log.SetOutput(writer)
-	log.SetPrefix(prefix + " ")
-	log.SetFlags(0)
-
-	return nil
-}
-
 // SetupLogger configures the logger to write to the kernel ring buffer via
 // /dev/kmsg.
-func SetupLogger(logger *log.Logger, prefix string, withLogFile bool) error {
+//
+// If logger is nil, default `log` logger is redirectred.
+//
+// If extraWriter is not nil, logs will be copied to it as well.
+func SetupLogger(logger *log.Logger, prefix string, extraWriter io.Writer) error {
```

Comment:
merged two methods together to reduce code duplication. 

now it's abstract enough (just accepts extraWriter), so it can deal with any logging subsystem

Reply:
Not gonna block on this, but does is `log.Logger` an `io.Writer`? If so, then what do you think about making this a variadic function:

```
func SetupLogger(prefix string, loggers ...io.Writer) error
```

I probably am way off, still need to look into this function some more.

Reply:
there are two issues with that:

* default logger (`log.Printf`) doesn't expose itself as a value we can pass in
* what we do is that we configure `logger`, and make it write to both `kmsg` and passed in `extraWriter`. we can make `extraWriters` variadic, but we can't mix `logger` and `extraWriters` as they are used in different ways

what we could do instead is that we could split this function into two functions:

1. function which returns writer to `/dev/kmsg` in this package
2. in some other package function which sets up logging the way we want for a list of writers (which will include in runtime kmsg and logging framework writers)

Reply:
The current implementation works. I will leave it up to you how for you wanna go with it. Something we could always circle back to.